From daaa7172e8f09e0165309bffa26a65372f402115 Mon Sep 17 00:00:00 2001
From: Jan Beich <jbeich@FreeBSD.org>
Date: Sat, 11 Jan 2020 21:28:37 +0000
Subject: [PATCH] Get framebuffer resolution on more platforms

media_driver/linux/common/codec/ddi/media_libva_decoder.cpp:48:10: fatal error: 'linux/fb.h' file not found
 #include <linux/fb.h>
          ^~~~~~~~~~~~
media_driver/linux/common/codec/ddi/media_libva_decoder.cpp:55:43: error: variable has incomplete type 'struct fb_var_screeninfo'
    struct fb_var_screeninfo              vsinfo;
                                          ^
media_driver/linux/common/codec/ddi/media_libva_decoder.cpp:62:22: error: use of undeclared identifier 'FBIOGET_VSCREENINFO'
        if(ioctl(fd, FBIOGET_VSCREENINFO, &vsinfo) < 0)
                     ^
media_driver/linux/common/ddi/media_libva.cpp:39:10: fatal error: 'linux/fb.h' file not found
 #include <linux/fb.h>
          ^~~~~~~~~~~~
---
 .../common/codec/ddi/media_libva_decoder.cpp  | 37 ++++++++++++++++++-
 media_driver/linux/common/ddi/media_libva.cpp |  2 -
 2 files changed, 35 insertions(+), 4 deletions(-)

diff --git a/media_driver/linux/common/codec/ddi/media_libva_decoder.cpp b/media_driver/linux/common/codec/ddi/media_libva_decoder.cpp
index f6dba1902f..d26deabd24 100755
--- a/media_driver/linux/common/codec/ddi/media_libva_decoder.cpp
+++ b/media_driver/linux/common/codec/ddi/media_libva_decoder.cpp
@@ -47,7 +47,40 @@
 #include <X11/Xutil.h>
 #endif
 
+#if defined(__linux__)
 #include <linux/fb.h>
+#define DEFAULT_FBDEV "/dev/graphics/fb0"
+#elif defined(__DragonFly__) || defined(__FreeBSD__) || defined(__FreeBSD_kernel__) || defined(__sun)
+#include <sys/fbio.h>
+# if defined(__sun)
+#define DEFAULT_FBDEV "/dev/fb"
+# else
+#define DEFAULT_FBDEV "/dev/ttyv0"
+# endif
+#define FBIOGET_VSCREENINFO FBIOGTYPE
+#define fb_var_screeninfo fbtype
+#define xres fb_width
+#define yres fb_height
+#elif defined(__NetBSD__) || defined(__OpenBSD__)
+#include <dev/wscons/wsconsio.h>
+# if defined(__OpenBSD__)
+#define DEFAULT_FBDEV "/dev/ttyC0"
+# else
+#define DEFAULT_FBDEV "/dev/ttyE0"
+# endif
+#define FBIOGET_VSCREENINFO WSDISPLAYIO_GINFO
+#define fb_var_screeninfo wsdisplay_fbinfo
+#define xres width
+#define yres height
+#else
+#warning "Cannot query framebuffer properties on this platform."
+#define DEFAULT_FBDEV "/dev/fb0"
+#define FBIOGET_VSCREENINFO 0
+struct fb_var_screeninfo {
+    uint32_t xres;
+    uint32_t yres;
+};
+#endif
 
 typedef MediaDdiFactory<DdiMediaDecode, DDI_DECODE_CONFIG_ATTR> DdiDecodeFactory;
 
@@ -64,8 +97,8 @@ static int32_t DdiDecode_GetDisplayInfo(VADriverContextP ctx)
     vsinfo.xres                           = 0;
     vsinfo.yres                           = 0;
 
-    fd = open("/dev/graphics/fb0",O_RDONLY);
-    if(fd >= 0)
+    fd = open(DEFAULT_FBDEV,O_RDONLY);
+    if(fd > 0)
     {
         if(ioctl(fd, FBIOGET_VSCREENINFO, &vsinfo) < 0)
         {
diff --git a/media_driver/linux/common/ddi/media_libva.cpp b/media_driver/linux/common/ddi/media_libva.cpp
index f52c6e65c2..f706b0cda2 100755
--- a/media_driver/linux/common/ddi/media_libva.cpp
+++ b/media_driver/linux/common/ddi/media_libva.cpp
@@ -36,8 +36,6 @@
 #include <X11/Xutil.h>
 #endif
 
-#include <linux/fb.h>
-
 #include "media_libva.h"
 
 #include "media_libva_util.h"
